src/s3/util/messages.ts

```diff
@@ -3,16 +3,69 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
-import { showLogOutputChannel } from '../../shared/logger'
+import * as vscode from 'vscode'
+import { getLogger, showLogOutputChannel } from '../../shared/logger'
 import { localize } from '../../shared/utilities/vsCodeUtils'
 import { Window } from '../../shared/vscode/window'
 
-export function showErrorWithLogs(message: string, window: Window): void {
+/**
+ * Shows a non-modal error message with a button to open the log output channel.
+ *
+ * @returns	A promise that resolves when the button is clicked or the error is dismissed.
+ */
+export async function showErrorWithLogs(message: string, window: Window): Promise<void> {
     const logsItem = localize('AWS.generic.message.viewLogs', 'View Logs...')
 
-    window.showErrorMessage(message, localize('AWS.generic.message.viewLogs', 'View Logs...')).then(selection => {
-        if (selection === logsItem) {
-            showLogOutputChannel()
-        }
-    })
+    return window
+        .showErrorMessage(message, localize('AWS.generic.message.viewLogs', 'View Logs...'))
+        .then(selection => {
+            if (selection === logsItem) {
+                showLogOutputChannel()
+            }
+        })
+}
+
+/**
+ * Shows a modal confirmation (warning) message with buttons to confirm or cancel.
+ *
+ * @param prompt the message to show.
+ * @param confirm the confirmation button text.
+ * @param cancel the cancel button text.
+ * @param window the window.
+ */
+export async function showConfirmationMessage(
+    { prompt, confirm, cancel }: { prompt: string; confirm: string; cancel: string },
+    window: Window
+): Promise<boolean> {
+    const confirmItem: vscode.MessageItem = { title: confirm }
+    const cancelItem: vscode.MessageItem = { title: cancel, isCloseAffordance: true }
+
+    const selection = await window.showWarningMessage(prompt, { modal: true }, confirmItem, cancelItem)
+    return selection === confirmItem
+}
+
+/**
+ * Creates a readable path to an s3 bucket or object (e.g. s3://...).
+ *
+ * This is the format used by AWS CLI.
+ * @see https://docs.aws.amazon.com/cli/latest/reference/s3/#path-argument-type
+ *
+ * @param bucket contains the name of the bucket.
+ * @param path to the object, or an empty string if this is the root of the bucket.
+ * @returns the readable path to the s3 bucket or object (e.g. s3://...).
+ */
+export function readablePath({ bucket, path }: { bucket: { name: string }; path: string }): string {
+    return path ? `s3://${bucket.name}/${path}` : `s3://${bucket.name}`
+}
+
+/**
+ * Shows an output channel and writes a line to it and also to the logs (as info).
+ *
+ * @param message the line to write.
+ * @param outputChannel the output channel to write to.
+ */
+export function showOutputMessage(message: string, outputChannel: vscode.OutputChannel) {
```

Comment:
I think this is the point of the ChannelLogger? That being said, this is a lot more elegant and I wouldn't mind seeing that go away.

Reply:
* Is there an equivalent method or ability to write an output message without force showing the output channel? That can lead to spammy behavior, and there are times when the toolkit would simply want to have text available if users are inclined to switch over and see it.
* consider adding an option to conditionally write to the logger. Not all output channel content needs to be logged, and perhaps some content contains account specific details which shouldn't be logged. Maybe this option should be off by default.

Reply:
> I think this is the point of the ChannelLogger? That being said, this is a lot more elegant and I wouldn't mind seeing that go away.

Yes, I saw that. Unfortunately it seems to make too many assumptions and I never converted this to the winston logger format. I do agree that this is such a simple function to implement that it doesn't necessarily warrant such an abstraction. That's basically what led me to just add the function rather than re-purposing the ChannelLogger?

> Is there an equivalent method or ability to write an output message without force showing the output channel? That can lead to spammy behavior, and there are times when the toolkit would simply want to have text available if users are inclined to switch over and see it.

> consider adding an option to conditionally write to the logger. Not all output channel content needs to be logged, and perhaps some content contains account specific details which shouldn't be logged. Maybe this option should be off by default.

I agree, but I'm gonna appeal to the "doesn't really need a general-purpose implementation because it's so simple" argument above. Might be more difficult to use the general purpose impl than to just implement the desired behavior, if you know what I mean (log level variations, difference in log message versus output channel message, showing channel, focusing channel, etc). It's not such much a pattern as it is a circumstantial UI choice.